### PR TITLE
Add TokenProvider implementation for client to request the token from the Azure Function

### PR DESCRIPTION
**Pull request by sonalivdeshpande**
This PR is an implementation of the fetching the FRS token from the Azure Function. The class FrsAzFunctionTokenProvider is responsible for fetching the token from the Azure Function URL mentioned using an axios-get call.

Initially, if the FLUID_CLIENT was FRS, the InsecureTokenProvider would generate a token on the client-side. 
Through this PR, if the FLUID_CLIENT is FRS, FrsAzFunctionTokenProvider is making a GET request to the Azure function 
by passing in the function url and the user credentials (either userId/userName or both). The userId and userName are 
part of the FrsAzFuncUser interface.

### CODE REVIEW

File: examples/hosts/app-integration/external-controller/src/app.ts [hunk 1]
```diff
@@ -12,11 +18,16 @@ import { renderAudience, renderDiceRoller } from "./view";
 // Define the server we will be using and initialize Fluid
 const useFrs = process.env.FLUID_CLIENT === "frs";
 
+const frsAzUser: FrsAzFuncUser = {
+    userId: "",
+    userName: "",
+};
+
 const user = generateUser();
```
Comment (sumedhb1995) [inline examples/hosts/app-integration/external-controller/src/app.ts:26]: Could we have the frsAzUser object be filled out by what is returned by generateUser?

File: experimental/framework/frs-client/src/interfaces.ts [hunk 2]
```diff
@@ -51,4 +51,9 @@ export interface FrsResources {
     containerServices: FrsContainerServices;
 }
 
+export interface FrsAzFuncUser {
```
Comment (sumedhb1995) [inline experimental/framework/frs-client/src/interfaces.ts:54]: Could we consider having this object be defined using `Pick` on the `FrsMember` object so we maintain a mapping between what's returned by the audience and what we pass into the `FrsAzFunctionTokenProvider`?

**Review by sumedhb1995 — COMMENTED** [review]
This is looking great! Thank you! Just a few comments regarding passing user details.
Also, either in this PR or a follow-up, could we update the external-controller README to reflect the changes?

File: experimental/framework/frs-client/src/FrsAzFunctionTokenProvider.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        };
+    }
+
+    public async fetchStorageToken(tenantId: string, documentId: string): Promise<ITokenResponse> {
+        return {
+            jwt: await this.getToken(tenantId, documentId),
+        };
+    }
+
+    private async getToken(tenantId, documentId: string): Promise<string> {
```
Comment (sumedhb1995) [inline experimental/framework/frs-client/src/FrsAzFunctionTokenProvider.ts:28]:
```suggestion
    private async getToken(tenantId: string, documentId: string): Promise<string> {
```

File: common/lib/protocol-definitions/src/users.ts [hunk 4]
```diff
@@ -8,4 +8,5 @@
  */
 export interface IUser {
     id: string;
+    name?: string;
```
Comment (vladsud) [inline common/lib/protocol-definitions/src/users.ts:11]: This protocol change seems like unrelated, or at least should be in separate PR?

Comment (sdeshpande3) [reply common/lib/protocol-definitions/src/users.ts:11]: Yep! It would be a part of separate PR. I will remove this protocol change from here.

File: experimental/framework/frs-client/README.md [hunk 5]
```diff
@@ -43,14 +43,14 @@ const frsClient = new FrsClient(config);
...
+When running against a live FRS instance, we can use the same interface as we do locally but instead using the tenant ID, orderer, and storage URLs that were provided as part of the FRS onboarding process. Each tenant ID maps to a tenant key secret that can be passed to the `FrsAzFunctionTokenProvider` to generate and sign the token such that the service will accept it. To ensure that the secret doesn't get exposed, it is passed to a secure, backend Azure function from which the token is fetched.
 
 ```typescript
 import { FrsClient, FrsConnectionConfig } from "@fluid-experimental/frs-client";
 
-const config: FrsConnectionConfig = { 
+const config: FrsConnectionConfig = {
     tenantId: "YOUR-TENANT-ID-HERE",
-    tokenProvider: new InsecureTokenProvider("YOUR-TENANT-ID-HERE" { id: "123", name: "Test User" }),
+    tokenProvider: new FrsAzFunctionTokenProvider("AZURE-FUNCTION-URL"+"/api/GetFrsToken", { userId: "test-user",userName: "Test User" }),
```
Comment (sumedhb1995) [inline experimental/framework/frs-client/README.md:53]: nit
```suggestion
    tokenProvider: new FrsAzFunctionTokenProvider("AZURE-FUNCTION-URL"+"/api/GetFrsToken", { userId: "test-user", userName: "Test User" }),
```

File: experimental/framework/frs-client/src/interfaces.ts [hunk 6]
```diff
@@ -44,6 +44,7 @@ export interface FrsContainerServices {
  */
 export interface FrsMember extends IMember {
     userName: string;
+    userId: string;
```
Comment (sumedhb1995) [inline experimental/framework/frs-client/src/interfaces.ts:47]: the `userId` field should already be available here from `IMember`

**Review by sumedhb1995 — APPROVED** [review]
This looking great! Thanks for setting up the integration between the Azure function service and the client!! Just a small comment on `FrsMember` typing.

A follow-up here to consider would be having the `user` object being passed into `FrsAzFunctionTokenProvider` also contain an additional field for user metadata that we could encode into the token as well when we make the request. And then parse it out in `FrsAudience`. This would allow the developer to include any additional information for the user that they will get returned in calls to the audience

File: examples/hosts/app-integration/external-controller/README.md [hunk 7]
```diff
@@ -42,3 +42,26 @@ For in browser testing update `./jest-puppeteer.config.js` to:
 Dice Roller uses the following distributed data structures:
 
 - SharedDirectory - root
+
+## Backed Locally and running with live FRS instance
+
+We can connect to both live FRS instance by passing in the tenanID, orderer and storage as well as using the tenantID as "local" for running against Tinylicious for development purpose.
+
```
Comment (sumedhb1995) [inline examples/hosts/app-integration/external-controller/README.md:49]: Could we add in a line here on how to launch the tinylicious service
```suggestion
"To launch the local Tinylicious service instance, run `npx tinylicious` from your terminal window"
```

File: examples/hosts/app-integration/external-controller/README.md [hunk 8]
```diff
@@ -42,3 +42,26 @@ For in browser testing update `./jest-puppeteer.config.js` to:
...
+For running the Tinylicious client, we pass the tenantID as "local" and make use of `InsecureTokenProvider` where we can pass anything into the key since we are running it locally and an object identifying the current user. For running the instance locally, it would get naviagted to the Tinylicious on the default values of `localhost:7070`.
```
Comment (sumedhb1995) [inline examples/hosts/app-integration/external-controller/README.md:50]:
```suggestion
To run the the `FrsClient` against our local tinylicious instance, we pass the `tenantId` as "local" and make use of `InsecureTokenProvider`. For the latter, we pass in two values to its constructor: a key string, which can be anything since we are running it locally, and an object identifying the current user. For running the instance locally, the orderer and storage URLs would point to the Tinylicious instance on the default values of `http://localhost:7070`.
```

File: examples/hosts/app-integration/external-controller/README.md [hunk 9]
```diff
@@ -42,3 +42,26 @@ For in browser testing update `./jest-puppeteer.config.js` to:
 Dice Roller uses the following distributed data structures:
 
 - SharedDirectory - root
+
+## Backed Locally and running with live FRS instance
+
+We can connect to both live FRS instance by passing in the tenanID, orderer and storage as well as using the tenantID as "local" for running against Tinylicious for development purpose.
```
Comment (sumedhb1995) [inline examples/hosts/app-integration/external-controller/README.md:48]: Let's have a uniform way of talking about "tenant ID". Whenever we refer to as the value itself, we should call it with the space, and if we are specifying it as the parameter in code, lets have it as `tenantId` in a code block

File: examples/hosts/app-integration/external-controller/README.md [hunk 10]
```diff
@@ -42,3 +42,26 @@ For in browser testing update `./jest-puppeteer.config.js` to:
...
+
+When running the live FRS Instance, we would require the tenant ID, orderer and storage URLs. We make use of `FrsAzFunctionTokenProvider` which takes in the Azure function URL and an object identifying the current user, thereby making an axios `GET` request call to the Azure Function. This axios call takes in the tenantID, documentId and userID/userName as optional parameters. The Azure Function is responsible for mapping the tenantId to tenant key secret to generate and sign the token such that the service will accept it.
+
+```typescript
+const connectionConfig: FrsConnectionConfig = useFrs ? {
+    tenantId: "YOUR-TENANT-ID-HERE",
+    tokenProvider: new FrsAzFunctionTokenProvider("AZURE-FUNCTION-URL"+"/api/GetFrsToken", { userId: "test-user", userName: "Test User" }),
+    orderer: "https://alfred.eus-1.canary.frs.azure.com",
+    storage: "https://historian.eus-1.canary.frs.azure.com",
```
Comment (sumedhb1995) [inline examples/hosts/app-integration/external-controller/README.md:59]:
```suggestion
    storage: "YOUR-STORAGE-URL-HERE",
```

File: examples/hosts/app-integration/external-controller/README.md [hunk 11]
```diff
@@ -42,3 +42,26 @@ For in browser testing update `./jest-puppeteer.config.js` to:
...
+
+For running the Tinylicious client, we pass the tenantID as "local" and make use of `InsecureTokenProvider` where we can pass anything into the key since we are running it locally and an object identifying the current user. For running the instance locally, it would get naviagted to the Tinylicious on the default values of `localhost:7070`.
+
+When running the live FRS Instance, we would require the tenant ID, orderer and storage URLs. We make use of `FrsAzFunctionTokenProvider` which takes in the Azure function URL and an object identifying the current user, thereby making an axios `GET` request call to the Azure Function. This axios call takes in the tenantID, documentId and userID/userName as optional parameters. The Azure Function is responsible for mapping the tenantId to tenant key secret to generate and sign the token such that the service will accept it.
+
+```typescript
+const connectionConfig: FrsConnectionConfig = useFrs ? {
+    tenantId: "YOUR-TENANT-ID-HERE",
+    tokenProvider: new FrsAzFunctionTokenProvider("AZURE-FUNCTION-URL"+"/api/GetFrsToken", { userId: "test-user", userName: "Test User" }),
+    orderer: "https://alfred.eus-1.canary.frs.azure.com",
```
Comment (sumedhb1995) [inline examples/hosts/app-integration/external-controller/README.md:58]:
```suggestion
    orderer: "YOUR-ORDERER-URL-HERE",
```

File: experimental/framework/frs-client/README.md [hunk 12]
```diff
@@ -43,16 +42,16 @@ const frsClient = new FrsClient(config);
...
 ```typescript
 import { FrsClient, FrsConnectionConfig } from "@fluid-experimental/frs-client";
 
-const config: FrsConnectionConfig = { 
-    tenantId: "YOUR-TENANT-ID-HERE",
-    tokenProvider: new InsecureTokenProvider("YOUR-TENANT-ID-HERE" { id: "123", name: "Test User" }),
-    orderer: "https://alfred.eus-1.canary.frs.azure.com",
-    storage: "https://historian.eus-1.canary.frs.azure.com",
+const config: FrsConnectionConfig = {
+    tenantId: "YOUR-TENANT-KEY-HERE",
```
Comment (sumedhb1995) [inline experimental/framework/frs-client/README.md:51]:
```suggestion
    tenantId: "YOUR-TENANT-ID-HERE",
```

File: examples/hosts/app-integration/external-controller/README.md [hunk 13]
```diff
@@ -42,3 +42,28 @@ For in browser testing update `./jest-puppeteer.config.js` to:
...
+
+"To launch the local Tinylicious service instance, run `npx tinylicious` from your terminal window"
+
+When running the live FRS Instance, we would require the tenant ID, orderer and storage URLs. We make use of `FrsAzFunctionTokenProvider` which takes in the Azure function URL and an object identifying the current user, thereby making an axios `GET` request call to the Azure Function. This axios call takes in the tenant ID, documentId and userID/userName as optional parameters. The Azure Function is responsible for mapping the tenantId to tenant key secret to generate and sign the token such that the service will accept it.
+
+```typescript
+const connectionConfig: FrsConnectionConfig = useFrs ? {
+    tenantId: "YOUR-TENANT-KEY-HERE",
```
Comment (sumedhb1995) [inline examples/hosts/app-integration/external-controller/README.md:58]:
```suggestion
    tenantId: "YOUR-TENANT-ID-HERE",
```
